### ISSUE TITLE
Time Series Support

### ISSUE THREAD

**Issue by adamjstewart** [torchgeo/torchgeo#2382]
This issue is designed to track progress towards full time series support in TorchGeo. If you are interested in helping with this effort, please join the `#time-series` channel on [Slack](https://join.slack.com/t/torchgeo/shared_invite/zt-22rse667m-eqtCeNW0yI000Tl4B~2PIw) and attend our weekly meetings.

#### Required changes

- [x] Replace rtree with geopandas: #2747
- [x] Replace fiona with geopandas: #2962
- [x] Spatial-only intersection: #2837
- [x] Xarray support: #509, #1490, #2945
- [x] Collation functions for padding: #2921
- [x] Convert change detection datasets: #2862, #2863, #2878, #2879, #2906, #3028
- [x] Return time series: #2308, #3183
- [ ] Convert time series datasets: #3042, #2921, #3368, #3369, #3371, #3372
- [ ] Splitters: guarantee non-overlapping spatial and/or temporal splits: @hkristen 

#### Desired changes

- [x] Replace BoundingBox with slicing: #2804
- [x] Support Polygon ROIs in all splitters: #2835
- [x] Deprecate BoundingBox: #2847
- [x] Store true image footprint: #2903
- [ ] Properly handle vector datasets: @isaaccorley
- [ ] Cache geopandas index: @isaaccorley
- [ ] STACDataset: see #stac channel
- [ ] ERA5
- [ ] CMIP6
- [ ] Support integer indexing?: @adamjstewart
- [ ] Unify GeoDataset and NonGeoDataset?: #1353
- [ ] Intersection/Union multiple queries?: #2408
- [ ] Tuple/MultiQuery dataset?: #877

#### Required changes

- [x] Support Polygon ROIs in all samplers: #2812
- [x] Separate spatial and temporal samplers: #3552
- [x] Ensure samplers do not sample outside Polygon bounds: #3552
- [x] Visualization: visualize an entire epoch with polygons and points: #3552

#### Desired changes

- [ ] Support weighted samplers?: #757
- [ ] Pre-computation?: run sample_points once per epoch for debugging?: #2300

### Transforms

- [x] B x T x C x H x W <-> B x TC x H x W: #2477 (TODO: image + mask)
- [ ] Temporal embeddings: #2826
- [ ] Temporal up/downsampling: https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.interpolate.html
- [ ] UTC time <-> local time
- [ ] Data imputation: https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.interpolate.html

#### 1D

- [x] L-TAE: #2867
- [x] Presto: #2929
- [x] Tessera: #3243 
- [ ] Seq2Seq: #2639 
- [ ] TempCNN
- [ ] Chronos v1, v2: @cnagda?
- [ ] Galileo: #3305

#### 2.5D

- [x] BTC: #2926
- [x] ChangeStar: #157 
- [x] ChangeViT: #2910
- [x] FC-Siam: #108, #345

#### 3D

- [x] Satlas: #1358, #1884, #2038
- [x] ConvLSTM: #2898
- [x] OlmoEarth: #3517
- [ ] CNN-LSTM: #2938
- [ ] AnySat
- [ ] OmniSat
- [ ] Prithvi v1, v2: @isaaccorley 
- [ ] SatMAE: @isaaccorley 
- [ ] SkySense v1, v2

#### 4D

- [x] Aurora: #2823
- [ ] ACE: #3402

### Tasks

- [ ] TC -> scalar, classification: TemporalClassification
- [x] TC -> scalar, regression: TemporalRegression (#3668)
- [x] 2CHW -> HW, classification: ChangeDetection (#2422, #2890)
- [ ] TCHW -> scalar, classification: SpatioTemporalClassification (#3591)
- [ ] TCHW -> scalar, regression: SpatioTemporalRegression (#3591)
- [x] TCHW -> HW, classification: SpatioTemporalSegmentation (#3671)
- [ ] TCHW -> HW, regression: SpatioTemporalPixelwiseRegression (#3086)
- [ ] TCHW -> BHW, classification: SpatioTemporalInstanceSegmentation
- [ ] TCZYX -> TCZYX, regression: WeatherForecasting

### Tutorials

- [ ] 1D: air pollution tutorial (#3669)
  - [x] Dataset: air quality (#3619)
  - [x] Model: L-TAE (#2867)
  - [x] Trainer: TemporalRegressionTask (#3668)
- [x] 2.5D: Change detection tutorial (#3222)
  - [x] Dataset: OSCD100 (#3221)
  - [x] Model: BTC (#2926)
  - [x] Trainer: ChangeDetectionTask (#2422)
- [ ] 3D: Crop type mapping tutorial (#3706)
  - [x] Dataset: PASTIS100 (#3265)
  - [x] Model: ConvLSTM (#2898)
  - [x] Trainer: SpatioTemporalSegmentationTask (#3671)
- [ ] 4D: Weather forecasting tutorial (#3654)
  - [ ] Dataset: WeatherBench (#3649)
  - [x] Model: Aurora (#2823)
  - [ ] Trainer: WeatherForecastingTask (not strictly required right now)

### Outdated discussions

- #640
- #1544

**Comment by DarthReca** [comment]
@adamjstewart, I will try to take a look at how to implement them this weekend since I have QuakeSet to test on

**Comment by gatienc** [comment]
@adamjstewart Is there any plan to support giving as output of the datamodule, not only the TxCxHxW Tensor but also the dates for each of the satellites acquisitions? this can become useful when the model do some positional embedding on the temporal aspect

**Comment by adamjstewart** [comment]
Definitely! Biggest hurdle we currently have is standardizing on a single key for this. Our datasets currently use `t`, `time`, `datetime`, `timestamp`, and many other possible names. Same thing for location, which could be `x`, `y`, `lat`, `lon`, `latitude`, `longitude`, etc.

Other annoyance is that torch.Tensor does not support datetime64 (unlike numpy) so we have to cast to a `pd.Timestamp().timestamp` float and cast back whenever we actually use it. I imagine different models will want the datetime in different formats.
